rust/sbp/src/de.rs

```diff
@@ -174,6 +178,153 @@ impl std::fmt::Display for CrcError {
+//         self.0.next()
+//     }
+// }
+
+pub struct Framer<R> {
+    inner: FramedRead<R, SbpFramer>,
+}
+
+impl<R: io::Read> Framer<R> {
+    pub fn new(reader: R) -> Self {
+        Self {
+            inner: FramedRead::new(reader, SbpFramer),
+        }
+    }
+}
+
+impl<R: io::Read> Iterator for Framer<R> {
+    type Item = Result<SbpFrame<BytesMut>, Error>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.inner.next()
+    }
+}
+
+// not sure a good naming for this atm
+pub struct SbpDecode<R>(FramedRead<R, SbpFramer>);
+
+impl<R: io::Read> SbpDecode<R> {
+    pub fn new(reader: R) -> Self {
+        SbpDecode(FramedRead::new(reader, SbpFramer))
+    }
+}
+
+impl<R: io::Read> Iterator for SbpDecode<R> {
+    type Item = Result<Sbp, Error>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.0.next().and_then(|frame| {
+            frame.map_or(None, |f| {
+                Some(
+                    f.check_crc()
+                        .map_err(|e| de::Error::from(e))
+                        .and_then(|_| Sbp::from_sbp_frame(f).map_err(|e| de::Error::from(e))),
+                )
+            })
+        })
+    }
+}
+
+#[derive(Debug, Clone, Default)]
+struct SbpFramer;
+
+impl Decoder for SbpFramer {
+    type Item = SbpFrame<BytesMut>;
+    type Error = Error;
+
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        Ok(SbpFrame::<BytesMut>::parse(src))
+    }
+    fn decode_eof(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        todo!()
+    }
+}
+
+#[derive(Debug)]
+pub struct SbpFrame<B>(B);
+
+impl<B: bytes::Buf> SbpFrame<B> {
+    pub fn msg_type(&self) -> u16 {
+        let mut slice = &self.0.chunk()[1..];
+        slice.get_u16_le()
+    }
+
+    pub fn sender_id(&self) -> u16 {
+        let mut slice = &self.0.chunk()[3..];
+        slice.get_u16_le()
+    }
+
+    pub fn payload_len(&self) -> usize {
+        self.0.chunk()[5] as usize
+    }
+
+    pub fn payload(&self) -> &[u8] {
+        &self.0.chunk()[HEADER_LEN..HEADER_LEN + self.payload_len()]
+    }
+
+    pub fn crc(&self) -> u16 {
+        let mut slice = &self.0.chunk()[HEADER_LEN + self.payload_len()..];
+        slice.get_u16_le()
+    }
+
+    pub fn check_crc(&self) -> Result<u16, CrcError> {
+        let actual = self.crc();
+        let data = &self.0.chunk()[1..HEADER_LEN + self.payload_len()];
+        if actual == crc16::State::<crc16::XMODEM>::calculate(data) {
+            Ok(actual)
+        } else {
+            Err(CrcError {
+                msg_type: self.msg_type(),
+                sender_id: self.sender_id(),
+                crc: actual,
+            })
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.0.chunk().len()
+    }
+}
+
+impl SbpFrame<BytesMut> {
+    pub fn parse(buf: &mut BytesMut) -> Option<Self> {
+        if let Some(index) = buf.iter().position(|b| b == &PREAMBLE) {
```

Comment:
Do we want to maybe wrap this in SbpFrame or just leave these skips as they are and return None?

Reply:
Actually I think we can skip on the parse functions for now and just do everything in the decoder

```rust
impl Decoder for SbpFramer {
    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
        let start = match src.iter().position(|b| b == &PREAMBLE) {
            Some(idx) => idx,
            None => {
                src.clear();
                return Ok(None);
            }
        };
        src.advance(start);
        if src.len() < HEADER_LEN {
            src.reserve(MAX_FRAME_LEN);
            return Ok(None);
        }
        let payload_len = src[5] as usize;
        let at = HEADER_LEN + payload_len + CRC_LEN;
        if src.len() < at {
            src.reserve(MAX_FRAME_LEN);
            return Ok(None);
        }
        let frame = SbpFrame(src.split_to(at));
        frame.check_crc()?;
        Ok(Some(frame))
    }
}
```

simplifies things not having to worry about parsing from `&[u8]` and we can add it later if we want it